### DISCUSSION TITLE
Enable and use homing on A  Axis using NC IR switch

### DISCUSSION THREAD

**Discussion by thegigi** [fra589/grbl-Mega-5X#244] [Q&A] [answered]
Hi Gauthier,

First  many thanks for the excellent port of grbl to 5 axis and wish you a happy new year 2022 !
Unfortunately i was not able to figure out how to enable and use homing on A (rotary) axis. On A axis i have a worm drive + stepper, a disc with hole and an optical endstop NC switch (https://www.aliexpress.com/item/32434841034.html?spm=a2g0s.9042311.0.0.27424c4dvAAF6C) connected  to D40 on Ramps board.  Hardware & software limits are not enabled. The other max endstops (X,Y,Z) are normally closed and work correctly by using the invert limit pins feature  (uncommenting line 282 in config.h)
//#define INVERT_MIN_LIMIT_PIN_MASK ((1<<AXIS_1) | (1<<AXIS_2) | (1<<AXIS_3))
Also how to use the A homing ? When i try $HA (from cncjs terminal)  i get no error but a completely erratic behavior of the grbl afterwards (even sw reset does not work anymore and i have to reconnect to grbl). 
Also do you think that it would be possible to use the signal from A axis endstop to get a kind of rpm reading for that axis ?

Andrei

**Comment by fra589** [comment]
Hi @thegigi,  
  
If you use the default axis configuration from Github, the A axis is the axis number 4.  
By default, the HOMING_CYCLE is only defined for X, Y and Z in config.h.  
You need to add additional axes to the right homing definition section (depending of N_AXIS you defined) between the lines 170 and the line 196 of config.h.  
  
@++;  
Gauthier.

**Reply by thegigi** [reply]
Hi Gauthier,

Thx for the swift answer . I have uncommented the lines :
#if N_AXIS == 4 // 4 axis : homing
  #define HOMING_CYCLE_0 (1<<AXIS_3) // Home Z axis first to clear workspace.
  //#define HOMING_CYCLE_1 ((1<<AXIS_1)|(1<<AXIS_2))     // OPTIONAL: uncomment to move X,Y at the same time.
  #define HOMING_CYCLE_1 (1<<AXIS_1) // Home X axis  // OPTIONAL: uncomment to move only X at a time.
  #define HOMING_CYCLE_2 (1<<AXIS_2) // Home Y axis  // OPTIONAL: uncomment to move only Y at a time.
  #define HOMING_CYCLE_3 (1<<AXIS_4) // Home 4th axis (A)
and uploaded the grbl .
Now the command $H does homing Z,X,Y but not A  and then grbl freezes. Command $HA also freezes grbl.
Is there any other change that i need to do to enable Axis A homing ?

Merci a+ :) 
Andrei

**Comment by fra589** [comment]
Hi Andrei,  

Your homing definition is fine.
Reading your first post, I see that you connect the A limit switch on D40, ie. on max limit for A axis.
So, since it's NC switch, you must define INVERT_MIN_LIMIT_PIN_MASK by uncomment the line 285 of config.h with the appropriate axis number:
`#define INVERT_MAX_LIMIT_PIN_MASK ((1<<AXIS_4))`
  
To check the operation of limit switch, you can see the result of real time "?" command. It shoult have a "Pn:A" field when the A limit switch is triggered.  
  
@++;  
Gauthier.

**Reply by thegigi** [reply]
Hi Gauthier,

Sry the switch works as normally open as there is a disc with a hole not a disc with a tab and operates correctly on D40. This is also the lathe spindle axis when driven by a DC motor (this the machine on FB Home Built CNC Machines Group - Andrei PS) 

Result of ? command is :
triggered : <Idle|WPos:139.371,-16.000,-2.240,-123.964|FS:0,0|Pn:A>
not triggered : <Idle|WPos:139.371,-16.000,-2.240,-123.964|FS:0,0>

Still the $H and $HA does not home A axis and instead brings the grbl to nirvana :). Is there any other adjustment to be done ?

Merci,
Andrei

**Reply by fra589** [reply]
> Result of ? command is :
> triggered : <Idle|WPos:139.371,-16.000,-2.240,-123.964|FS:0,0|Pn:A>
> not triggered : <Idle|WPos:139.371,-16.000,-2.240,-123.964|FS:0,0>

This mean that your A limit switch is well functioning...
I need to make hardware test to verify if I can reproduce your behavior...

**Reply by fra589** [reply]
Can you post the result of $I and $$ Grbl commands? So I will can try to reproduce with same parameters...

**Reply by thegigi** [reply]
Hi Gauthier,

thank you for your swift answer :
CNCjs 1.9.23 [Grbl]
Connected to /dev/ttyUSB0 with a baud rate of 19200
Grbl 1.2d ['$' for help]
client> $$
[MSG:'$H'|'$X' to unlock]
$0=10 (Step pulse time, microseconds)
$1=254 (Step idle delay, milliseconds)
$2=0 (Step pulse invert, mask)
$3=0 (Step direction invert, mask)
$4=0 (Invert step enable pin, boolean)
$5=0 (Invert limit pins, boolean)
$6=0 (Invert probe pin, boolean)
$10=0 (Status report options, mask)
$11=0.020 (Junction deviation, millimeters)
$12=0.002 (Arc tolerance, millimeters)
$13=0 (Report in inches, boolean)
$20=0 (Soft limits enable, boolean)
$21=1 (Hard limits enable, boolean)
$22=1 (Homing cycle enable, boolean)
$23=0 (Homing direction invert, mask)
$24=50.000 (Homing locate feed rate, mm/min)
$25=100.000 (Homing search seek rate, mm/min)
$26=250 (Homing switch debounce delay, milliseconds)
$27=1.000 (Homing switch pull-off distance, millimeters)
$30=12000 (Maximum spindle speed, RPM)
$31=550 (Minimum spindle speed, RPM)
$32=0 (Laser-mode enable, boolean)
$100=3200.000 (X-axis travel resolution, step/mm)
$101=3200.000 (Y-axis travel resolution, step/mm)
$102=3200.000 (Z-axis travel resolution, step/mm)
$103=800.000
$110=100.000 (X-axis maximum rate, mm/min)
$111=200.000 (Y-axis maximum rate, mm/min)
$112=50.000 (Z-axis maximum rate, mm/min)
$113=3200.000
$120=5.000 (X-axis acceleration, mm/sec^2)
$121=5.000 (Y-axis acceleration, mm/sec^2)
$122=5.000 (Z-axis acceleration, mm/sec^2)
$123=5.000
$130=120.000 (X-axis maximum travel, millimeters)
$131=60.000 (Y-axis maximum travel, millimeters)
$132=45.000 (Z-axis maximum travel, millimeters)
$133=100000.000
ok
client> $X
[MSG:Caution: Unlocked]
ok
> $i
[VER:1.2d.20210930:]
[AXS:4:XYZA]
[OPT:VNMGH,35,255,24]
ok
>

**Reply by fra589** [reply]
Thanks, I will do test soon...  
  
But at first sight, here is a parameter that seems out of scope to me...  
The maximum speed of your A axis ($113=3200 (°/mn) seems abnormally high to me! This requires too high a frequency to reach this speed. This will cause the calculation capacities of the Arduino board to be exceeded and will cause it to crash.  
  
Does your A axis move correctly when bypassing the homing?  
  
Perhaps you should try to lower this speed around 350~400 for testing...  
  
@++;  
Gauthier.

**Reply by thegigi** [reply]
Bonjour  Gauthier,

Oui tu a raison $113=3200 units/min c'est trop.  J-ai mis $113=400 mais le comportement au sujet du $HA est identique. 
A ajouter qu'au moment du $HA le moteur A produit un bruit tres court similaire a celui du reset.

aussi j-ai teste avec 2000 :
> $113=2000
ok
> g01g91A300f1000
ok
et ca fonctionne ok. 

Avec g01g91A300f2000 il  donne un glitch au moment de vitesse maximale et s-arrete mais le grbl reste usable.

Mais avec $ha et $h ca donne  un crash. Chez toi le $ha ca fonctionne ?

Merci,
Andrei

**Comment by fra589** [comment, marked as answer]
Bonjour Andrei,  
  
Tu parles Français, c'est plus facile pour moi ! 😃  
  
Chez moi, le $HA fonctionne...  
Pendant que j'y pense,... N'aurais-tu pas par hasard testé d'autres version de grbl sur ta cartes Mega ou d'autres configuration d'axes ?  
Si tel est le cas, as tu réinitialisé les "paramètres d'usine" avec la commande grbl: `$RST=*` ?  
  
En cas de changement de configuration d'axes (nombre, noms / clonage, etc.) la mémoire flash sauvegardant les paramètres ne correspond plus à la définition des axes et cela peut produire des résultats imprévisibles. C'est même la raison principale des derniers problèmes remontés sur Github.  
  
@++;  
Gauthier.

**Reply by thegigi** [reply]
Super ... Merci Gauthier !!.. avec flash reset  : $RST=* et recharge des parametres le $H et le $HA et la machine fonctionnent comme prevu. Maintenant je peux passer au travaux plus complexes.

Encore une fois merci et bonne chance comme developeur et gardien de ce logiciel !!

Andrei

**Comment by ecncshopInd** [comment]
Iam trying to use the Grbl for mega firmware. 
I need to do homing of my theta axis. 
I tried enabling "#define HOMING_SINGLE_AXIS_COMMANDS " in config.h
Then Iam able to home the X,Y,Z axis individually using $HX, $HY, $HZ commands.
But $HA is not working. Is it supported in grbl for mega ?

Some of the answers above are not in English. I will try google translate, but let me know if any one has tried this recently.

**Reply by RaphaelDives** [reply]
Hey,
what do you expect $HA to do? I use it for homing my A-Axis... That works perfect!
I guess, you want to Home-All? But isn't that the purpose of the $H - homing cycle?
Cheers,
Raphael

**Reply by fra589** [reply]
Hi @ecncshopInd,  
  
> But $HA is not working.  
  
It's hard to answer with so little information!  
  
The simple fact of saying "is not working" generally does not make it possible to find the origin of a malfunction.  
 When 'it doesn't work', is there an error message? any particular behavior or noise? a Grbl crash? anything else?  
Which version, configuration and what settings are you using? Post the results of the $I and $$ commands...  
Which changes have you made in config.h? Normally, you had to adapt and configure the parameters HOMING_CYCLE_0 -> HOMING_CYCLE_N...
Are your limits switches NO, NC or mix? Perhaps you need to configure INVERT_MIN_LIMIT_PIN_MASK...  
  
@++;  
Gauthier.

**Reply by ecncshopInd** [reply]
Thanks for ur replies.
Actually Iam using mega from this link: https://github.com/dguerizec/grbl-Mega-4axis
My version is showing 1.1e on bootup. 
Is there any better 4-axis grbl version available ?

I have enabled individual axis homing in config.h using 
#define HOMING_SINGLE_AXIS_COMMANDS

Now $HX $HY $HZ are working. But when I give $HA, I get error code 3 ! (unsupported grbl commmand ). Iam able to jog , move theta axis. But not able to Home. 

Even in the config.h comments, $HA is not mentioned as a supported command. 

But $HA seems to be working for some of u. Can u clarify which version of grbl for mega are u using. Is there anything Iam missing.

**Reply by fra589** [reply]
Hi @ecncshopInd,  
  
> Actually Iam using mega from this link: https://github.com/dguerizec/grbl-Mega-4axis  
  
Here, it's the original place for [grbl-Mega-5X](https://github.com/fra589/grbl-Mega-5X), not dguerizec/grbl-Mega-4axis.  
You will only get answers corresponding to grbl-Mega-5X, I don't know if it's the same thing in the other versions that I don't maintain!  
  
$HA work in grbl-Mega-5X.

@++;  
Gauthier.

**Comment by RaphaelDives** [comment]
Hey,
I was using it I think some versions ago… but I‘m sure I did nothing special for it. It worked out of the box… I will look it up by chance.

Just to make sure: the option for single axis homing is enabled by default. What version did you use? The grbl mega4 that you mention or the grbl-mega5x from this page? 

Maybe installing the current version (5x of course) and setting the parameters new can help? (And don’t forget the reset command… 😉)

Still you could also post the axis definition and the $$ output. 😉

Cheers
Raphael
